### ISSUE TITLE
/usr/local/lib/ocaml/version:3:3: error: invalid preprocessing directive

### ISSUE THREAD

**Issue by Sjord** [semgrep/semgrep#3158]
I try to build semgrep-core from git sources on macOS 11.3.1. I get errors:

```
$ make
rm -f bin
./scripts/make-version > ./src/cli/version.ml
dune build
       clang src/tree-sitter-lang/semgrep-cpp/lib/scanner.o (exit 1)
(cd _build/default/src/tree-sitter-lang/semgrep-cpp/lib && /usr/bin/clang -fPIC -I . -g -I /usr/local/lib/ocaml -I /Users/sjoerdlangkemper/.opam/default/lib/ANSITerminal -I /Users/sjoerdlangkemper/.opam/default/lib/atdgen-runtime -I /Users/sjoerdlangkemper/.opam/default/lib/base/caml -I /Users/sjoerdlangkemper/.opam/default/lib/biniou -I /Users/sjoerdlangkemper/.opam/default/lib/bytes -I /Users/sjoerdlangkemper/.opam/default/lib/cmdliner -I /Users/sjoerdlangkemper/.opam/default/lib/containers -I /Users/sjoerdlangkemper/.opam/default/lib/containers/monomorphic -I /Users/sjoerdlangkemper/.opam/default/lib/easy-format -I /Users/sjoerdlangkemper/.opam/default/lib/parsexp -I /Users/sjoerdlangkemper/.opam/default/lib/ppx_deriving/runtime -I /Users/sjoerdlangkemper/.opam/default/lib/ppx_sexp_conv/runtime-lib -I /Users/sjoerdlangkemper/.opam/default/lib/result -I /Users/sjoerdlangkemper/.opam/default/lib/seq -I /Users/sjoerdlangkemper/.opam/default/lib/sexplib -I /Users/sjoerdlangkemper/.opam/default/lib/sexplib0 -I /Users/sjoerdlangkemper/.opam/default/lib/tsort -I /Users/sjoerdlangkemper/.opam/default/lib/yojson -I ../../../ocaml-tree-sitter/src/bindings/lib -I ../../../ocaml-tree-sitter/src/gen/lib -I ../../../ocaml-tree-sitter/src/run/lib -o scanner.o -c scanner.cc)
In file included from scanner.cc:2:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string:506:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string_view:175:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__string:57:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/algorithm:639:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/initializer_list:46:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/cstddef:37:
/usr/local/lib/ocaml/version:1:1: error: expected unqualified-id
4.10.0
^
/usr/local/lib/ocaml/version:3:3: error: invalid preprocessing directive
# The version string is the first line of this file.
  ^
/usr/local/lib/ocaml/version:4:3: error: invalid preprocessing directive
# It must be in the format described in stdlib/sys.mli
  ^
In file included from scanner.cc:2:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string:506:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string_view:175:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__string:57:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/algorithm:639:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/initializer_list:46:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/cstddef:49:9: error: no member named 'ptrdiff_t' in the global namespace
using ::ptrdiff_t;
      ~~^
In file included from scanner.cc:2:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string:506:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/string_view:175:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__string:57:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/algorithm:640:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:418:
/usr/local/lib/ocaml/version:1:1: error: expected unqualified-id
4.10.0
^
```

And more similar errors. It seems that the include path is incorrect and ocaml/version is used instead of another version.cc. Temporarily moving ocaml/version (with `sudo mv /usr/local/lib/ocaml/version /usr/local/lib/ocaml/version2`) solves the problem.

**Comment by brendongo** [comment]
Hi @Sjord thanks for filling! 

This is due to https://github.com/ocaml/ocaml/issues/9523 and was fixed in Ocaml 4.12 https://github.com/ocaml/ocaml/commit/b727f422bd48022a830566f19c7d5e50181a96e8 we are in the process of updating to 4.12 (https://github.com/returntocorp/semgrep/issues/3027) 

Right now we are also using the workaround you used by moving/removing the offending version file :D

**Comment by aryx** [comment]
yes it's because C++ brilliantly now use header files without even a .h suffix ... so any version file in a directory in the include PATH will be used. Thanks C++ ...
